### PR TITLE
Item swap between hands hook

### PR DESCRIPTION
**Pull request by nflantier**
Added a hook to handle the sucess or not of swapping item when pressing the swap item key.

### CODE REVIEW

File: patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch [hunk 1]
```diff
@@ -1,6 +1,37 @@
 --- ../src-base/minecraft/net/minecraft/network/NetHandlerPlayServer.java
 +++ ../src-work/minecraft/net/minecraft/network/NetHandlerPlayServer.java
-@@ -611,7 +611,10 @@
+@@ -90,7 +90,9 @@
+ import net.minecraft.tileentity.TileEntityCommandBlock;
+ import net.minecraft.tileentity.TileEntitySign;
+ import net.minecraft.tileentity.TileEntityStructure;
++import net.minecraft.util.ActionResult;
+ import net.minecraft.util.ChatAllowedCharacters;
++import net.minecraft.util.EnumActionResult;
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch]: Don't change imports in patches.

File: patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch [hunk 2]
```diff
@@ -1,6 +1,37 @@
...
++                    ItemStack itemstack2 = this.field_147369_b.func_184586_b(EnumHand.MAIN_HAND);
++                    
++                    ActionResult<ItemStack> res1 = itemstack1 != null ? itemstack1.func_77973_b().onSwapHand(this.field_147369_b, EnumHand.OFF_HAND, itemstack1, EnumHand.MAIN_HAND, itemstack2) : new ActionResult(EnumActionResult.PASS, null);
++                    ActionResult<ItemStack> res2 = itemstack2 != null ? itemstack2.func_77973_b().onSwapHand(this.field_147369_b, EnumHand.MAIN_HAND, itemstack2, EnumHand.OFF_HAND, itemstack1) : new ActionResult(EnumActionResult.PASS, null);
++                    
++                    if(res1.func_188397_a() == EnumActionResult.PASS && res2.func_188397_a() == EnumActionResult.PASS)
++                    {
++                    	this.field_147369_b.func_184611_a(EnumHand.OFF_HAND, res2.func_188398_b());
++                    	this.field_147369_b.func_184611_a(EnumHand.MAIN_HAND, res1.func_188398_b());
++                    }
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch]: This whole ordeal can be simplified to a 2 line patch. Externalize all the checks into ForgeHooks and don't change the indentation on vanilla code.

File: patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch [hunk 3]
```diff
@@ -1,6 +1,22 @@
...
+                 {
+-                    ItemStack itemstack1 = this.field_147369_b.func_184586_b(EnumHand.OFF_HAND);
+-                    this.field_147369_b.func_184611_a(EnumHand.OFF_HAND, this.field_147369_b.func_184586_b(EnumHand.MAIN_HAND));
+-                    this.field_147369_b.func_184611_a(EnumHand.MAIN_HAND, itemstack1);
++                	List<ItemStack> l = net.minecraftforge.common.ForgeHooks.onSwapHand(this.field_147369_b);
++                	if(!l.isEmpty())
++                	{
++                		this.field_147369_b.func_184611_a(EnumHand.OFF_HAND, l.get(0));
++                    	this.field_147369_b.func_184611_a(EnumHand.MAIN_HAND, l.get(1));
++                	}
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch]: Not what I meant.
Leave the lines that actually swap the stacks as is (no indentation, nothing). Then add `if FH.onSwapHand(...) {` before it and `}` after it. No need for any `List` return or anything.

**Comment by nflantier** [comment]
I need the list return because the user can specify the stack he wants in the ActionResult, its not just a condition checking.

**Comment by diesieben07** [comment]
Uh, yeah don't do that then. If the mod wants to change the stack they can do that themselves after denying the vanilla swap. Just make it a boolean return.

File: patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch [hunk 4]
```diff
@@ -1,6 +1,20 @@
 --- ../src-base/minecraft/net/minecraft/network/NetHandlerPlayServer.java
 +++ ../src-work/minecraft/net/minecraft/network/NetHandlerPlayServer.java
-@@ -611,7 +611,10 @@
+@@ -571,9 +571,12 @@
+ 
+                 if (!this.field_147369_b.func_175149_v())
+                 {
+-                    ItemStack itemstack1 = this.field_147369_b.func_184586_b(EnumHand.OFF_HAND);
++                	if(net.minecraftforge.common.ForgeHooks.onSwapHand(this.field_147369_b))
++                	{
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/network/NetHandlerPlayServer.java.patch]: Tabs -.-

**Comment by LexManos** [comment]
Why? What does it matter if the hands are swapped?

**Comment by nflantier** [comment]
i have some item that need to stay in their slot while they have a certain nbt data

**Comment by LexManos** [comment]
What do you mean by that? How are these items used? WHY do they need to stay in their slot? Why not just remove the NBT when they are swapped? Or when you detect they are in a different slot.

**Comment by nflantier** [comment]
when the player right click with the item it set a boolean nbt to true and spawn an entity and pass the slot number of the item in the constructor, when that entity is dead the item s nbt boolean is set to false by the slot number. Sure i tried to detect when they are in diffrent slot it works fine execpt when your item is in the 0 slot of your main inventory because onupdate just provide the slot number and you get the slot 0 from your main inventory and a 0 from your offhand inventory too, so you cant detect the change. Maybe changing onupdate to have the type of inventory the item is in is the good solution.

**Comment by LexManos** [comment]
Why not just send the itemstack?

**Comment by nflantier** [comment]
because in the entity constructor when i assign the itemstack passed in param to the itemstack params of the class it will become a new instance of that item stack and no longer a reference so it will not be linked to the player inventory

**Comment by diesieben07** [comment]
You cannot possibly track all movement of items in the player's inventory. One proper way to do this would be to store the entity's UUID in the ItemStack and then search for that in the inventory.

**Comment by LexManos** [comment]
> because in the entity constructor when i assign the itemstack passed in param to the itemstack params of the class it will become a new instance of that item stack and no longer a reference so it will not be linked to the player inventory

Assigning something does not create a new instance...

But either way as Dies said, there are ways to do what you want and still allow swapping. It would be a nightmare to try and figure out and cancel ALL the ways you could possibly move a item in an inventory....

**Comment by nflantier** [comment]
its not creating a new instance but the itemstack of the class after assigning the itemstack passed in parameters is no longer a pointer to the itemstack contained in the array representing the inventory of the player.

**Comment by LexManos** [comment]
Then you are doing something wrong because assigning doesnt make new instances. Java is all byref for objects.

**Comment by diesieben07** [comment]
You are throwing many words around, some of which accurate, some of which are not.
Even if the ItemStack would still be a "pointer" to the player inventory, this would break horribly as soon as the player logs out, switches dimensions, etc.

This is no longer a topic for a github issue though.

**Comment by LexManos** [comment]
Closing as this is not well articulated, and from what we can understand, what he wants to do is not what this PR is actually doing. And he can do it better/cleaner without this PR using functions that already exist today.

**Pull request closed without merging**